### DISCUSSION TITLE
MessageOutput improvements for OpenDtu-onBattery?

### DISCUSSION THREAD

**Discussion by MalteSchm** [tbnobody/OpenDTU#1130] [General]
Hi,

I thought I'd ask for guidance here.

in the OpenDTU-onBattery project we use the MessageOutput class to output information from the features added in the -onBattery project. This works soso and I believe it interferes with debug messages that originate from Hoymiles interrupts. A random example copied from the console is shown below. In this example the printout of the PowerLimiter is interrupted by the CTX RealTimeRunData message.

I was wondering if there is some guidance here how this situation can be improved? Can we print messages in another way? Can we/I contribute code to OpenDTU that solves this. If so what would be the preferred approach to do this? 

```
23:31:41.452 > [PowerLimiterClass::loop] newPowerLimit: 190
23:31:41.452 > [PowerLimiterClass::setNewPowerLimit] 4 channels total, 4 producing channels, scaling power limit
23:31:41.452 > [PowerLimiterCTX RealTimeRunData Channel: 75 --> 15 83 07 46 82 80 14 75 68 80 0B 00 64 A8 84 3C 00 00 00 00 00 00 00 00 DB BF 47
```

**Comment by tbnobody** [comment]
I would guess that the message from the power limiter gets printed from within a callback routine which runs in a different thread? In this case it would be required to interlock the print routine(s) with a semaphore. But this on the other hand would block callbacks / interrupt routines / normal methods which are maybe time critical. It could even lead to dead-locks.

**Reply by MalteSchm** [reply]
Hi @tbnobody 

The Power Limiter messages gets printed from a method which is called from the main loop. I somehow guessed that the Hoymiles related messages are send within an interrupt but I don't know this part of the code.... so I may be wrong.
If this is the same thread then this behavior confuses me...

I agree that locking seems not ideal. I wondered if one of these options could be a solution too:
a) an API to set which information gets logged. I understand that it may not be of much use for the OpenDTU project. But I could then implement a Radio Button to specify which log messages should be output
b) put the OpenDTU Log messages under a define. This could be enabled for OpenDTU and we/users could disable this if debugging of non Hoymiles communication related issues is required

**Reply by schlimmchen** [reply]
Please ignore this comment, as it was ill-informed. See #1212 for an update.

> But this on the other hand would block callbacks / interrupt routines / normal methods which are maybe time critical.

~~This translates to: "OpenDTU prints messages using MessageOutput from within interrupt contexts." I translate it like that because as far as I can tell OpenDTU has no notion of threads. That's why only interrupt contexts are able to interrupt other pieces of code. Am I wrong?~~

~~MessageOutput should only be used outside of interrupts. If an interrupt changes a particular state, the respective message should be printed later when the change is detected in the context of the regular program flow.~~

~~Having said that my suggestion would be to return a wrapper in `Hoymiles::getMessageOutput()` that buffers messages. The buffered messages are only sent to the global `MessageOutput` instance within the `Hoymiles::loop()` function. Does that make sense? Is this too memory-intensive?~~

**Comment by schlimmchen** [comment]
Related: #1212.
